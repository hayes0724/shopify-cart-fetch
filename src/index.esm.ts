/**
 * Cart
 * ------------------------------------------------------------------------------
 * The default Cart class with all methods, events and state
 */

import CartEvents from "./lib/events";
import {
  addItem,
  addItemFromForm,
  clearAttributes,
  clearItems,
  clearNote,
  getAttributes,
  getNote,
  getState,
  removeItem,
  updateAttributes,
  updateItem,
  updateItemById,
  updateNote,
} from "./lib/functions";
import {
  CartItemParams,
  CartItemAdd,
  CartItemRemove,
  //CartItemUpdate,
  CartItemIndex,
  CartLineItem,
  CartState,
  EventType,
  LineItemKey,
  LineItemCallback,
  CartItemUpdate,
  CartItemUpdateById,
  Attributes,
  CartParams,
} from "./lib/types";

/**
 * This class implements all methods from functions.ts, uses events.ts for cart events and maintains cart state.
 * You can also import the functions individually or create your own cart class.
 *
 * @example
 * import Cart from '@hayes0724/shopify-cart-fetch';
 * const shopifyCart = new Cart();
 * shopifyCart.getState().then(state => console.log(state))
 * @example
 * import {getState, removeItem, updateItem, addItem, clearItems} from '@hayes0724/shopify-cart-fetch'
 * getState().then(state => console.log(state))
 *
 */
export default class Cart {
  get state(): CartState {
    return this._state;
  }

  set state(value: CartState) {
    this._state = value;
  }

  private _state?: CartState;

  private async run(
    type: EventType,
    callback: CallableFunction,
    data?: CartParams|CartItemParams
  ): Promise<CartState> {
    this.events.emit(type, "start");
    this.state = await callback(data);
    this.events.emit(type, "complete", this.state);
    return this.state;
  }

  private async runLine(
    type: EventType,
    callback: LineItemCallback,
    data?: CartItemParams | CartItemParams[]
  ): Promise<CartLineItem | CartLineItem[]> {
    this.events.emit(type, "start");
    const lineItem = await callback(data);
    const index = this.findItemIndex(lineItem.key);
    index > -1 && lineItem.quantity > 0
      ? (this.state.items[index] = lineItem)
      : this.state.items.push(lineItem);
    this.events.emit(type, "complete", this.state);
    return lineItem;
  }

  public events = new CartEvents();

  /**
   * Finds the the line item using it's line item key
   * @param {LineItemKey} key
   * @return {CartLineItem}
   */
  public findItem = (key: LineItemKey): CartLineItem =>
    this.state.items.find((item) => item.key === key);

  /**
   * Finds the index of the line item using it's line item key
   * @param {LineItemKey} key
   * @return {CartItemIndex}
   */
  public findItemIndex = (key: LineItemKey): CartItemIndex =>
    this.state.items.findIndex((item) => item.key === key);

  public getState = async (): Promise<CartState> =>
    this.run("cart-fetch", getState);

  public addItem = async (items: CartItemAdd[]): Promise<CartLineItem> =>
    this.runLine("cart-add", addItem, items);

  public addItemFromForm = async (
    productForm: HTMLFormElement
  ): Promise<CartLineItem> =>
    this.runLine("cart-add", addItemFromForm, productForm);

  public clearItems = async (): Promise<CartState> =>
    this.run("cart-update", clearItems);

  public removeItem = async (item: CartItemRemove): Promise<CartState> =>
    this.run("cart-remove", removeItem, item);

  public updateItem = async (item: CartItemUpdate): Promise<CartState> =>
    this.run("cart-update", updateItem, item);

  public updateItemById = async (
    item: CartItemUpdateById
  ): Promise<CartLineItem> => this.runLine("cart-update", updateItemById, item);

  public updateAttributes = async (
    attributes: Attributes
  ): Promise<CartState> =>
    this.run("cart-update", updateAttributes, attributes);

  public clearAttributes = async (): Promise<CartState> =>
    this.run("cart-update", clearAttributes);

  public getAttributes: () => Promise<CartState["attributes"]> = getAttributes;

  public getNote: () => Promise<CartState["note"]> = getNote;

  public updateNote = async (note: string): Promise<CartState> =>
    this.run("cart-update", updateNote, note);

  public clearNote = async (): Promise<CartState> =>
    this.run("cart-update", clearNote);
}

export {
  addItem,
  getState,
  addItemFromForm,
  clearItems,
  removeItem,
  updateAttributes,
  updateItem,
  updateItemById,
  getAttributes,
  clearAttributes,
  getNote,
  updateNote,
  clearNote
};

export type {
  CartItemParams,
  CartItemAdd,
  CartItemRemove,
  CartItemIndex,
  CartLineItem,
  CartState,
  EventType,
  LineItemKey,
  LineItemCallback,
  CartItemUpdate,
  CartItemUpdateById,
  Attributes,
  CartParams
};

