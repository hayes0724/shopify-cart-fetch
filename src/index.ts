export * from './shopify-cart';
export * from './lib/types';
export * from './lib/cart-interface';
export * from './lib/errors';
