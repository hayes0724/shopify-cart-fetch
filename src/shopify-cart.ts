import {CartInterface} from "./lib/cart-interface";
import {
  Attributes, CartEvents,
  CartItemRemove,
  CartItems,
  CartItemsResponse, CartItemUpdate,
  CartLineItem, CartRoute,
  CartSettings,
  CartState, ShopifyResponse
} from "./lib/types";
import {InventoryError, VariantError} from "./lib/errors";

export class ShopifyCart implements CartInterface {
  private _state: CartState;
  private readonly _settings: CartSettings = {
    url: '',
    postConfig: {
      method: 'POST',
      credentials: 'same-origin',
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
        'Content-Type': 'application/json;',
      },
    },
    updateState: true,
  };

  constructor(settings: CartSettings) {
    this._settings = { ...this._settings, ...settings };
    this.cartEvent('cart:ready');
  }

  get state(): CartState {
    return this._state;
  }

  get settings(): CartSettings {
    return this._settings;
  }

  /**
   * Use the GET /cart.js endpoint to get the cart as JSON.
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#get-cart-js | ShopifyAPI: GET /cart.js }
   */
  public async getState(): Promise<CartState> {
    this.cartEvent('cart:requestStarted', '/cart.js');
    const response = await fetch(`${this._settings.url}/cart.js`);
    this._state = await response.json();
    this.cartEvent('cart:requestComplete', '/cart.js');
    return this._state;
  }

  /**
   * Use the POST /cart/add.js endpoint to add product to cart.
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-add-js | ShopifyAPI: POST /cart/add.js}
   */
  public async addItem(items: CartItems): Promise<CartItemsResponse> {
    const data = Array.isArray(items) ? items : [items];
    const response = await this.post('/cart/add.js', JSON.stringify({ items: data }));
    if (this._settings.updateState) {
      await this.getState();
    }
    return response as CartItemsResponse;
  }

  /**
   * Adds an item to your cart from a product form. The form must contain an input with name="id".
   * If the quantity specified is more than what is available, the promise will be rejected and the cart state will remain unchanged
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-add-js | ShopifyAPI: POST /cart/add.js}
   */
  public async addItemFromForm(productForm: HTMLFormElement): Promise<CartLineItem> {
    const formData = new FormData(productForm);
    if (!formData.get('id')) {
      throw 'Cart form missing required property ID';
    }
    const formJson = JSON.stringify(Object.fromEntries(formData.entries()));
    const response = await this.post('/cart/add.js', formJson);
    if (this._settings.updateState) {
      await this.getState();
    }
    return response as CartLineItem;
  }

  /**
   * Clear all cart attributes from Shopify and return the state
   * @see {@link https://shopify.dev/api/ajax/reference/cart#post-cart-update-js | ShopifyAPI: POST /cart/update.js}
   * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes | ShopifyAPI: cart-attributes }
   */
  public async clearAttributes(): Promise<CartState> {
    const state = await this.getState();
    const data = JSON.stringify({
      attributes: this.clearProps(state.attributes),
    });
    return await this.post('/cart/update.js', data);
  }

  /**
   * Use the POST /cart/clear.js endpoint to set all quantities of all line items in the cart to zero.
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-clear-js | ShopifyAPI: POST /cart/clear.js}
   */
  public async clearItems(): Promise<CartState> {
    return await this.post('/cart/clear.js');
  }

  /**
   * Remove the cart note
   * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-note | ShopifyAPI: cart.note }
   */
  public async clearNote(): Promise<CartState> {
    return await this.post('/cart/update.js', JSON.stringify({ note: '' }));
  }

  /**
   * Removes an item from the cart using line item key or product id. Returns a promise which fulfills with the updated cart state.
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-change-js | ShopifyAPI: POST /cart/change.js }
   */
  public async removeItem(item: CartItemRemove): Promise<CartState> {
    return await this.post('/cart/change.js', JSON.stringify({ quantity: 0, ...item }));
  }

  /**
   * Update cart attributes from Shopify and return the state
   * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes | ShopifyAPI: cart-attributes }
   */
  public async updateAttributes(attributes: Attributes): Promise<CartState> {
    return await this.post('/cart/update.js', JSON.stringify({ attributes: { ...attributes } }));
  }

  /**
   * The /cart/change.js endpoint changes the quantity and properties object of a cart line item.
   * Only items already in your cart can be changed, and only one line item at a time can be changed.
   * @see {@link https://shopify.dev/docs/themes/ajax-api/reference/cart#post-cart-change-js | ShopifyAPI: POST /cart/change.js }
   */
  public async updateItem(item: CartItemUpdate): Promise<CartState> {
    return await this.post('/cart/change.js', JSON.stringify(item));
  }

  /**
   * Update or create the cart note
   * @see {@link https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-note | ShopifyAPI: cart.note }
   */
  public async updateNote(note: string): Promise<CartState> {
    return await this.post('/cart/update.js', JSON.stringify({ note: note }));
  }

  protected async post<Return>(route: CartRoute, data?: BodyInit): Promise<Return> {
    const url = this._settings.url + route;
    const postConfig = this._settings.postConfig;
    if (data) {
      postConfig.body = data;
    }
    this.cartEvent('cart:requestStarted', route);
    const request = await fetch(url, postConfig);
    const response = await request.json();
    this.checkResponse(response);
    this.cartEvent('cart:requestComplete', route);
    return response;
  }

  protected checkResponse(response: ShopifyResponse): void {
    if (response['status'] === 404) {
      throw new VariantError();
    }
    if (response['status'] === 422) {
      throw new InventoryError(response['description']);
    }
    if (response['token']) {
      this._state = response as CartState;
    }
  }

  protected clearProps(target: Record<string, unknown>): { [key: string]: '' } {
    const clearList = {};
    Object.getOwnPropertyNames(target).forEach((prop) => (clearList[prop] = ''));
    return clearList;
  }

  protected cartEvent(name: CartEvents, route?: CartRoute): void {
    document.dispatchEvent(
      new CustomEvent(name, {
        detail: {
          cart: this,
          route: route,
        },
        bubbles: true,
        cancelable: true,
        composed: false,
      })
    );
  }
}
