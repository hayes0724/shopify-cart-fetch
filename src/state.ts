/**
 * The current cart object from Shopify
 * https://shopify.dev/docs/themes/ajax-api/reference/cart#get-cart-js
 */
export class CartState {
  public token?: string;
  public note?: string;
  public attributes?: Attributes;
  public total_price?: number;
  public total_weight?: number;
  public item_count?: number;
  public items?: Items[];
  public requires_shipping?: boolean;
  public currency?: string;
  public items_subtotal_price?: number;
  public cart_level_discount_applications?: CartLevelDiscountApplications[];
}

/**
 * Shopify cart attributes
 * https://shopify.dev/docs/themes/liquid/reference/objects/cart#cart-attributes
 */
export class Attributes {
  [index: string]: string;
}

/**
 * Cart line_item object
 * https://shopify.dev/docs/themes/liquid/reference/objects/line_item/
 */
export class Items {
  public id?: number;
  public quantity?: number;
  public variant_id?: number;
  public key?: string;
  public title?: string;
  public price?: number;
  public line_price?: number;
  public final_price?: number;
  public final_line_price?: number;
  public sku?: string;
  public grams?: number;
  public vendor?: string;
  public taxable?: boolean;
  public product_id?: number;
  public product_has_only_default_variant?: boolean;
  public gift_card?: boolean;
  public url?: string;
  public featured_image?: FeaturedImage;
  public image?: string;
  public handle?: string;
  public requires_shipping?: boolean;
  public product_title?: string;
  public product_description?: string;
  public product_type?: string;
  public variant_title?: string;
  public variant_options?: string[];
  public options_with_values?: OptionsWithValues[];
}

export class FeaturedImage {
  public url?: string;
  public aspect_ratio?: number;
  public alt?: string;
}

export class OptionsWithValues {
  public name?: string;
  public value?: string;
}

export class CartLevelDiscountApplications {
  public type?: string;
  public key?: string;
  public title?: string;
  public description?: string;
  public value?: string;
  public created_at?: string;
  public value_type?: string;
  public allocation_method?: string;
  public target_selection?: string;
  public target_type?: string;
  public total_allocated_amount?: number;
}
