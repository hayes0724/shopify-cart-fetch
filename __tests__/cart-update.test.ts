import { enableFetchMocks } from "jest-fetch-mock";
enableFetchMocks();

import { ShopifyCart } from "../src";
import {
  cartResponse,
  settings,
  updateAttributesInput,
} from "./fixtures";

let cart: ShopifyCart;

const validate = (response) => {
  expect(cart.state).toEqual(response);
  expect(fetchMock.mock.calls.length).toEqual(1);
  expect(fetchMock.mock.calls[0][1]["body"]).toMatchSnapshot();
  expect(fetchMock.mock.calls[0][0]).toEqual(`${cart.settings.url}/cart/update.js`);
};

describe("Testing methods using the /cart/update.js endpoint", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    cart = new ShopifyCart(settings);
    fetchMock.mockResponse(JSON.stringify(cartResponse));
  });

  it("updateAttributes calls fetch and the correct default url", async () => {
    const response = await cart.updateAttributes(updateAttributesInput);
    validate(response);
  });

  it("clearAttributes calls fetch and the correct default url", async () => {
    const response = await cart.clearAttributes();
    expect(cart.state).toEqual(response);
    expect(fetchMock.mock.calls.length).toEqual(2);
    expect(fetchMock.mock.calls[0][0]).toEqual(`${cart.settings.url}/cart.js`);
    expect(fetchMock.mock.calls[1][0]).toEqual(`${cart.settings.url}/cart/update.js`);
    expect(fetchMock.mock.calls[1][1]["body"]).toMatchSnapshot();
  });

  it("updateNote calls fetch and the correct default url", async () => {
    const response = await cart.updateNote("note");
    validate(response);
  });

  it("clearNote calls fetch and the correct default url", async () => {
    const response = await cart.clearNote();
    validate(response);
  });
});
