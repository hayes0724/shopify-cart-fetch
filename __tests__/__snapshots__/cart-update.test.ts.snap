// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing methods using the /cart/update.js endpoint clearAttributes calls fetch and the correct default url 1`] = `"{\\"attributes\\":{\\"Gift wrap\\":\\"\\"}}"`;

exports[`Testing methods using the /cart/update.js endpoint clearNote calls fetch and the correct default url 1`] = `"{\\"note\\":\\"\\"}"`;

exports[`Testing methods using the /cart/update.js endpoint updateAttributes calls fetch and the correct default url 1`] = `"{\\"attributes\\":{\\"attribute\\":\\"value\\",\\"gift wrap\\":\\"true\\"}}"`;

exports[`Testing methods using the /cart/update.js endpoint updateNote calls fetch and the correct default url 1`] = `"{\\"note\\":\\"note\\"}"`;
